generic/MAPL_ServiceServices.F90

```diff
@@ -0,0 +1,176 @@
+#include "MAPL_ErrLog.h"
+module mapl_ServiceServices
+  use ESMF
+  use MAPL_ExceptionHandling
+  use mapl_ServiceConnectionItemVector
+  use mapl_ProvidedServiceItemVector
+  use mapl_RequestedServiceItemVector
+  use mapl_VariableSpecification
+
+  implicit none
+  private
+
+
+  public ProvidedServiceType
+  public ServiceConnectionGet
+  public ProvidedServiceGet
+  public ProvidedServiceSet
+  public RequestedServiceGet
+  public FillRequestBundle
+
+  interface ProvidedServiceType
+     module procedure newProvidedServiceType
+  end interface ProvidedServiceType
+
+contains
+
+  function newProvidedServiceType(SERVICE, BUNDLE) result (item)
+    character (len=*)             , intent(IN   ) :: SERVICE
+    character (len=*)             , intent(IN   ) :: BUNDLE
+    
+    type (ProvidedServiceType)  :: item
+            
+    item%Service_name = SERVICE
+    item%Bundle_name = BUNDLE
+  end function newProvidedServiceType
+
+  subroutine ServiceConnectionGet(item, &
+            Service, Provider, Requester, RC)
+
+    type (ServiceConnectionType), intent(IN)  :: ITEM
+    character (len=*), optional, intent(  OUT) :: SERVICE
+    character (len=*), optional, intent(  OUT) :: PROVIDER
+    character (len=*), optional, intent(  OUT) :: REQUESTER
+    integer,           optional, intent(  OUT) :: RC     ! Error code:
+    
+    if (present(Service)) then
+       Service = item%Service_name
+    end if
+
+    if (present(Provider)) then
+       Provider = item%Provider_name
+    end if
+
+    if (present(requester)) then
+       requester = item%requester_name
+    end if
+
+    _RETURN(ESMF_SUCCESS)
+  end subroutine ServiceConnectionGet
+
+  subroutine ProvidedServiceGet(provider_list, advertised_service, bundle, rc)
+    type(ProvidedServiceItemVector), intent(IN) :: provider_list
+    character(len=*), intent(IN) :: advertised_service
+    type(ESMF_FieldBundle), intent(OUT) :: bundle
+    integer, optional, intent(out) :: rc
+    
+    logical :: found
+    type (ProvidedServiceItemVectorIterator) :: iter
+    type (ProvidedServiceType), pointer :: item
+    
+    _ASSERT(provider_list%size()>0,'provider_list should not be empty')
+    
+    found = .false.
+    iter = provider_list%begin()
+    ! loop over provided services
+    do while (iter /= provider_list%end())
+       item => iter%get()
+       if(item%service_name == advertised_service) then
+          found = .true.
+          bundle = item%bundle
+          exit
```

Comment:
Let's leave this for now, but wouldn't a MAP work better than a vector here?   you could then eliminate this procedure entirely:

```f90
bundle => list%at(advertised_service)
```